lti/views.py

```diff
@@ -0,0 +1,183 @@
+import random
+import string
+
+from django.http import HttpResponseRedirect
+from django.shortcuts import render
+from django.urls import reverse
+from django.views.decorators.csrf import csrf_exempt
+from django.views.generic import RedirectView
+from django.views.generic.edit import FormView
+
+from django.contrib.auth import get_user_model, login
+
+from account.models import EmailAddress
+from lti_provider.lti import LTI
+from pylti.common import LTIException
+
+from groups.models import Group
+
+from .forms import LtiUsernameForm
+from .utils import login_existing_user
+
+
+LTI_PROPERTY_COURSE_ID = "custom_canvas_course_id"
+LTI_PROPERTY_COURSE_TITLE = "context_title"
+LTI_PROPERTY_USER_EMAIL = "lis_person_contact_email_primary"
+LTI_PROPERTY_USER_ROLES = "ext_roles"
+
+
+def get_random_alphanumeric_string(length):
+    letters_and_digits = string.ascii_letters + string.digits
+    result_str = "".join((random.choice(letters_and_digits) for i in range(length)))
+    return result_str
+
+
+class LtiInitializerView(RedirectView):
+
+    url = "/"
+
+    @csrf_exempt
+    def dispatch(self, request, *args, **kwargs):
+        """ Handle LTI verification and user authentication """
+        lti = LTI(request_type="any", role_type="any")
+        try:
+            lti.verify(request)
+        except LTIException:
+            return render(request, "lti_failure.html")
+
+        if not request.user.is_authenticated:
+            try:
+                lti_user = login_existing_user(request)
+            except EmailAddress.DoesNotExist:
+                request.session["lti_email"] = request.POST.get(LTI_PROPERTY_USER_EMAIL, None)
+                if request.session["lti_email"] is None:
+                    return render(request, "lti_failure.html")
+                return HttpResponseRedirect(reverse("lti_registration"))
+            if lti_user is False:
+                return render(request, "lti_failure.html")
+
+        return super(LtiInitializerView, self).dispatch(request, *args, **kwargs)
+
+    def get(self, request, *args, **kwargs):
+        """ Handle the redirection coming from username registration """
+
+        lti_params = {
+            "course_id": request.session.get(LTI_PROPERTY_COURSE_ID, None),
+            "roles": request.session.get(LTI_PROPERTY_USER_ROLES, None),
+            "title": request.session.get(LTI_PROPERTY_COURSE_TITLE, None)
+        }
+
+        if None in lti_params.values():
+            return render(request, "lti_failure.html")
+
+        self.initialize_group(
```

Comment:
It's generally bad practice to mutate anything on `GET` requests.  Is this necessary to do here?  It's not clear to me reading through things what the goal of calling `GET /lti/lti_initializer/` goal is or why it can't be a `POST` if it needs to create/change data.

Reply:
Good question. It was to handle the `success_url` redirect from the username registration form. I've never attempted to specify that redirection come in the form of `POST`, so I was just grabbing the lti properties that were set in the session by `django-lti-provider` in order to initialize the user's groups and roles.

Since Hedera moved to using a display_name, there is no need for a username registration form now, so this all is getting deleted anyways.